Staging Sites: Navigate away from the domains page when on a staging site

jeroenpf opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to: https://github.com/Automattic/dotcom-forge/issues/3048

When a user lands on the domains search page but the active site is a staging site, an error will be shown because staging sites do not have the option to manage domains.

## Proposed Changes

When a user lands on the search domains site while the current site is a staging site, navigate to the dashboard/home.

## Testing Instructions

1. Create a Business site
2. Activate hosting features and add a staging site
3. Navigate to Upgrades -> Domains on the production site
4. Click 'Add a domain' and then 'Search for a domain'
5. Search for the domain
6. Switch context to the staging site in the left sidebar selector
7. Confirm that you navigate to the homepage and not to the domains search page.

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

wojtekn commented:
@jeroenpf We've used a custom middleware to achieve similar behavior under https://github.com/Automattic/wp-calypso/pull/74257

I took a look into that PR, and I noticed we did it for domains too: 
https://github.com/Automattic/wp-calypso/blob/a8d2f9633d1b0c5c144ed3d9fb2abdaeb67ee770/client/my-sites/controller.js#L709

Would it be possible to handle this on a middleware level for domain search results, too? May it be that it stopped working at some point?

jeroenpf commented:
@wojtekn The middleware you mentioned is working and being triggered when loading the page or switching the site, but the state hasn't been updated yet so when the middleware is executed, it still sees the previous site. I will see if there is something we can do here.

jeroenpf committed:
Move staging redirect check down the middleware chain

The staging redirect check was being called too early, leading to issues. Moved it below another critical middleware it depended on to ensure correct processing order.

jeroenpf commented:
In the end, it turns out the middleware was being called too early in the chain. It was executed before the site change was processed. I moved it down. Now it's working. @wojtekn can you check if it's ok now?

wojtekn reviewed (approved):
@jeroenpf nice find, it works as expected now!
